build-system/pr-check/build-targets.js

```diff
@@ -45,10 +53,12 @@ const Targets = {
   DEV_DASHBOARD: 'DEV_DASHBOARD',
   DOCS: 'DOCS',
   E2E_TEST: 'E2E_TEST',
-  FLAG_CONFIG: 'FLAG_CONFIG',
   INTEGRATION_TEST: 'INTEGRATION_TEST',
+  LINT: 'LINT',
   OWNERS: 'OWNERS',
   PACKAGE_UPGRADE: 'PACKAGE_UPGRADE',
+  PRESUBMIT: 'PRESUBMIT',
+  PRETTIFY: 'PRETTIFY',
```

Comment:
_[Replying to top-level comment inline to enable comment threading]_

> For instance Targets seems to be an enum of name => String(name). Could this instead be a Map? Map<name, { matchPredicate, ... }>

I did this for two reasons:
1. So that logging a Set of enum variables that represent matched targets will print out their string representations.

https://github.com/ampproject/amphtml/blob/72a145352ba54627fa034a36009e125da4056740/build-system/pr-check/build-targets.js#L272-L275

2. So that CI scripts can check if a PR affects any of the given list of Targets

https://github.com/ampproject/amphtml/blob/72a145352ba54627fa034a36009e125da4056740/build-system/pr-check/checks.js#L57-L95

Do you think there's an easier way to accomplish both these things?

Reply:
I'm not certain it would be more aesthetically pleasing, but all three are doable.

The composition of the items could be...
1. Determine Targets
2. Leverage Target information

```
class Target {
  name: string;
  predicateFileMatch: (files: Array<string>) => boolean;
  cliExecutionCommand: Array<string>;
}
```

During the build target phase...
```
const targets = new Set();
for (const target of allTargets) {
  if (target.predicateFileMatch(files)) {
    targets.add(target);
  }
}

module.exports = {
  targets,
}
```

During "checks":
```
const {targets} = require("active-targets.js");
for (const target of targets) {
  for (const cliCommand of target.cliCommands) {
    timedExecOrDie(cliCommand);
  }
}
```

Reply:
Thanks for the suggestion. At first glance, I'm not sure this approach will work well for all our use cases.

Today, `build-targets.js` merely determines which targets are touched by a PR, while the various CI job scripts in `build-system/pr-check` decide for themselves what to do with that information.  It looks like the approach being proposed will transfer the second item to the `Target` class, which I'm afraid will make the code harder to read and maintain.

For example, several months ago, `@wg-caching` wanted Validator tests to be run when the `RUNTIME` target was affected, but `@wg-runtime` did not want Runtime tests to be run when the `VALIDATOR` target was affected.

Today, this logic is encapsulated in the CI job scripts in an easily readable state. E.g. _"If this PR touches runtime files or validator files, run the validator tests"_

https://github.com/ampproject/amphtml/blob/72a145352ba54627fa034a36009e125da4056740/build-system/pr-check/validator-tests.js#L48-L50

Okay to punt this rewrite to another day, in case I can find a way to make it work in a more readable / maintainable way?